### PR TITLE
MGDAPI-5010 implement gcp redis metrics

### PR DESCRIPTION
**Pull request by cecobask**
## Overview

Jira: https://issues.redhat.com/browse/MGDAPI-5010

## Verification

- Provision a GCP cluster
  - From the master branch of the [Delorean](https://github.com/integr8ly/delorean) repo create a file in the root of the repo called gcp_service_account.json. Request GCP credentials and add these to the file
  - Run:
    ```
    make -f make/ocm.mk ocm/cluster.json BYOC=true CLOUD_PROVIDER=gcp OCM_CLUSTER_REGION=europe-west2
    ```
  - Edit the cluster.json file to update the name (and remove the expiration timestamp if necessary). You can also reduce the number of nodes to 2
  - Run:
    ```
    make ocm/cluster/create
    ```
- Create a Redis instance
  - Checkout my pull request
    ```
    gh pr checkout 595
    ```
  - Log in to your cluster and run:
    ```
    PROVIDER=gcp make cluster/prepare
    ```
  - Install the operator from my index image:
    ```
    cat << EOF | oc create -f - -n openshift-marketplace
      apiVersion: operators.coreos.com/v1alpha1
      kind: CatalogSource
      metadata:
        name: cro
        namespace: openshift-marketplace
      spec:
        image: 'quay.io/tdimov0/cloud-resource-operator:index-v0.42.0'
        sourceType: grpc
    EOF
    ```
  - Create a Redis instance and wait until it provisions successfully:
    ```
    PROVIDER=gcp make cluster/seed/redis
    ```
  - Delete the Redis instance and wait for it to complete:
    ```
    oc delete redis example-redis -n cloud-resource-operator
    ```
  - Open a remote shell to the operator pod:
    ```
    PODNAME=$(oc get pods -n cloud-resource-operator -o json | jq -r '.items[0].metadata.name')
    oc rsh $PODNAME
    ```
  - Check if the `cro_***` metrics are available and showing accurate values:
    ```
    curl localhost:8383/metrics | grep cro_
    ```
  - You should get similar result to this:
    ```
    cro_redis_available{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",strategy="gcp-memorystore"} 1
    cro_redis_connection{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",strategy="gcp-memorystore"} 0
    cro_redis_info{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",status="DELETING",strategy="gcp-memorystore"} 1.6717962442590702e+09
    cro_redis_info{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",status="READY",strategy="gcp-memorystore"} 1.671796153324022e+09
    cro_redis_info{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",status="UPDATING",strategy="gcp-memorystore"} 1.6717956732873619e+09
    cro_redis_status_phase{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",statusPhase="complete",strategy="gcp-memorystore"} 0
    cro_redis_status_phase{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",statusPhase="deletion in progress",strategy="gcp-memorystore"} 1
    cro_redis_status_phase{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",statusPhase="failed",strategy="gcp-memorystore"} 0
    cro_redis_status_phase{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",statusPhase="in progress",strategy="gcp-memorystore"} 0
    cro_redis_status_phase{clusterID="tdimov-ccs-4q6mx",instanceID="tdimovccs4q6mxcloudresourceoperator-uhja",namespace="cloud-resource-operator",productName="productName",resourceID="example-redis",statusPhase="paused",strategy="gcp-memorystore"} 0
    ```

### CODE REVIEW

**Review by Boomatang — COMMENTED** [review]
I have only done a code review as I don't currently have gcp access. Also have ran the unit tests localy with all passing.

The comments I made about the function names ending in List if returning the list version of a type might be a larger topic which needs to be discussed with the wider team.

File: pkg/providers/gcp/provider_postgres.go [hunk 1]
```diff
@@ -123,26 +122,16 @@
...
 		return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
 	}
-	if address == nil || address.GetStatus() == computepb.Address_RESERVING.String() {
-		return nil, croType.StatusMessage("network ip address range creation in progress"), nil
-	}
-	logger.Infof("created ip address range %s: %s/%d", address.GetName(), address.GetAddress(), address.GetPrefixLength())
-
-	logger.Infof("creating network service connection")
-	service, err := networkManager.CreateNetworkService(ctx)
+	_, err = networkManager.CreateNetworkService(ctx)
```
[the remark this answers is not in the document]
Comment (cecobask) [reply pkg/providers/gcp/provider_postgres.go:130]: After refactoring the way `networkManager.CreateNetworkService()` works, the service struct was no longer needed. Here is where we ensure the function doesn't return until the service was created or we encountered an error: [func (n *NetworkProvider) waitForConnection(clusterVpc *computepb.Network) (*servicenetworking.Connection, error)](https://github.com/cecobask/cloud-resource-operator/blob/7888a185c36154ed8fac27bdbd9b54503b3dbdb6/pkg/providers/gcp/cluster_network_provider.go#L204:L233)

File: pkg/providers/gcp/provider_postgres.go [hunk 2]
```diff
@@ -123,26 +122,16 @@
 		errMsg := "failed to reconcile network provider config"
 		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
 	}
-	address, err := networkManager.CreateNetworkIpRange(ctx, ipRangeCidr)
+	_, err = networkManager.CreateNetworkIpRange(ctx, ipRangeCidr)
```
[the remark this answers is not in the document]
Comment (cecobask) [reply pkg/providers/gcp/provider_postgres.go:125]: After refactoring the way `networkManager.CreateNetworkIpRange()` works, the address struct was no longer needed. Here is where we ensure the function doesn't return until the address was created or we encountered an error: [func (n *NetworkProvider) waitForAddress(ctx context.Context, ipRangeName string) (*computepb.Address, error)](https://github.com/cecobask/cloud-resource-operator/blob/7888a185c36154ed8fac27bdbd9b54503b3dbdb6/pkg/providers/gcp/cluster_network_provider.go#L133:L162)

**Comment by cecobask** [comment]
@Boomatang, thank you for the review. I've addressed your feedback in the 7aa2970 commit.
Let me know if everything is okay now.

**Comment by Fiona-Waters** [comment]
The functionality works as expected when following the verification steps. 
Screenshot from 2023-01-11 14-42-41
@Boomatang are you happy with the code changes following your feedback? if yes then I'm happy to place lgtm.

**Comment by Boomatang** [comment]
@Fiona-Waters yes, I am happy with the changes, as you did the verifications I'll let you have place the lgtm

**Comment by austincunningham** [comment]
@cecobask why was there a need to make changes in the aws provider ? I would have taught that these changes would only be in the gcp provider

**Comment by cecobask** [comment]
@austincunningham, the changes are as a result of refactoring the prometheus metrics functions to be made more generic and allow reuse for any provider since the logic is identical. Also, I renamed some tagging functions, as suggested by @Boomatang, to be make it more obvious what their purpose is.

**Pull request merged**